go/src/url_shortener/url_shortener.go

```diff
@@ -0,0 +1,90 @@
+package main
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"log"
+	"net/http"
+	"sync"
+)
+
+var (
+	cache URLShortener = URLShortener{shorts: make(map[string]*urlInfo)}
+)
+
+// URLShortener URL shortener server data structure
+type URLShortener struct {
+	mux    sync.Mutex
+	shorts map[string]*urlInfo
+}
+
+type urlInfo struct {
+	longURL string
+
+	// statistics
+	shortURL string
+	//count    int
+}
+
+func (cache *URLShortener) addURL(longURL string, shortURL string) {
+	cache.mux.Lock()
+	defer cache.mux.Unlock()
+
+	cache.shorts[shortURL] = &urlInfo{longURL, shortURL}
+}
+
+func (cache *URLShortener) getURL(shortURL string) (string, error) {
+	cache.mux.Lock()
+	defer cache.mux.Unlock()
+
+	if info, ok := cache.shorts[shortURL]; ok {
+		return info.longURL, nil
+	}
+
+	return "", fmt.Errorf("short URL not found: %s", shortURL)
+}
+
+func shorten(longURL string) string {
+	hasher := sha1.New()
+
+	hasher.Write([]byte(longURL))
+	sum := hasher.Sum(nil)
+
+	shortURL := fmt.Sprintf("%x", sum)[:6]
+
+	return shortURL
+}
+
+func shortenerHandler(w http.ResponseWriter, r *http.Request) {
+	longURL := r.URL.Path[len("/shorten/"):]
```

Comment:
This handler could be in another situation, what would happen if it was not anymore routed from `/shorten` ? :-)
For string operations -> https://golang.org/pkg/strings/
For path-related operations -> https://golang.org/pkg/path/